[WIP] fix #13404 poEvalCommand now works on windows

timotheecour opened the pull request:
* fix #13404 poEvalCommand now works on windows
* added runnableExample
* `execCmdEx("echo ok1 && echo ok2")` now works on posix and windows

timotheecour committed:
fix #13404 poEvalCommand now works on windows

Araq commented on lib/pure/osproc.nim:
```diff
@@ -70,6 +70,10 @@ const poDemon* {.deprecated.} = poDaemon ## Nim versions before 0.20
                                          ## and this is needed just for backward compatibility.
 
+const useShPath {.strdefine.} =
+  when defined(windows): "sh"
```
"sh" is not a common Windows tool.

timotheecour replied:
fixed => now using `cmd /c`

Araq commented:
Completely unacceptable approach, so now ``poEvalCommand`` for Windows depends on some Unix emulation layer?

timotheecour commented:
* then what's `sh` best equivalent on windows, such that it would allow for example running 2 commands consecutively, eg:

 `execCmdEx("git checkout hotfix && nimble test")` ?

(not looking for a workaround like `doAssert execCmdEx("git checkout hotfix") == 0; doAssert execCmdEx("nimble test") == 0;` )

* running via sh / bash seems useful though, and would help cross platform code.

I could either add a param `sh` to `execProcess` + friends, eg:
`proc execProcess(command: string, ..., sh = "")` 
(where sh becomes `useShPath` if empty, and `useShPath` defaults to empty on windows)

or I could add a windows specific flag `poEvalCommandViaSh` (but that seems less flexible, eg not overrideable as bash / sh)

The logic in this PR is non-trivial (see `quoteShellWindowsForSh` which is different from `quoteShellWindows`) so implementing it in properly in stdlib once instead of letting the user stumble upon it each time makes sense

Clyybber commented:
Use cmd ?

timotheecour commented:
> Use cmd ?

@Clyybber can you please provide a full example? how would I run 2 commands using `cmd.exe`, eg:
```nim
import osproc
echo execProcess("sh -c 'echo ok1 && echo ok2'") # ok, prints: ok1\nok2\n (on my windows machine)
echo execProcess("cmd -c 'echo ok1 && echo ok2'") # this hangs
```
for a concrete example, `git checkout hotfix && nimble test`

Clyybber commented:
That is not the problem of poEvalCommand. It should just pass it on to the shell.
In this case use `cmd /c "..."`

timotheecour committed:
windows: use `cmd /c command` instead of `sh -c command.quoteShellWindowsForSh`

Clyybber commented on lib/pure/os.nim:
```diff
@@ -965,6 +965,37 @@ proc quoteShellWindows*(s: string): string {.noSideEffect, rtl, extern: "nosp$1"
     result.add("\"")
 
+when defined(windows):
+  proc quoteShellWindowsForSh*(s: string): string {.noSideEffect, rtl, extern: "nosp$1".} =
```
Remove this dead code please

timotheecour replied:
I've removed it to unblock this PR but it is useful in case user wants to use non-standard sh or bash instead of standard (but fraught with limitations) cmd.exe

timotheecour committed:
removed quoteShellWindowsForSh from this PR (even though was useful in case user wants to use non-standard but useful sh instead of cmd)

timotheecour commented:
> Completely unacceptable approach, so now poEvalCommand for Windows depends on some Unix emulation layer?

fixed; `cmd /c echo ok1 && echo ok2` does works, so I've updated PR to use `cmd /c` instead of `sh -c` on windows.
so now, `execCmdEx("echo ok1 && echo ok2")` does work cross-platform (and doesn't rely on non-standard tools)

Clyybber commented on lib/pure/osproc.nim:
```diff
@@ -70,6 +70,10 @@ const poDemon* {.deprecated.} = poDaemon ## Nim versions before 0.20
+const useShPath {.strdefine.} =
+  when defined(windows): "cmd"
+  elif not defined(android): "/bin/sh"
+  else: "/system/bin/sh"
```
Nit, but please do
```
elif defined(android): "/system/bin/sh"
else: "/bin/sh"
```
instead

timotheecour replied:
was pre-existing, but good point, fixed

timotheecour committed:
fix nit

Araq commented:
> (not looking for a workaround like doAssert execCmdEx("git checkout hotfix") == 0; doAssert execCmdEx("nimble test") == 0; )

That's not a "workaround", it's the only way to do this properly. Otherwise you're in shell quoting hell with all its non-portability and security nightmares.

timotheecour commented:
> That's not a "workaround", it's the only way to do this properly. Otherwise you're in shell quoting hell with all its non-portability and security nightmares.

I know, but sometimes it's inconvenient to modify third party code that was written with single command in mind just to support chaining multiple commands. Example: 
in important_packages.nim, if you want to apply a hotfix branch to a package that's broken (and the author is unresponsive or would be inconvenient to merge the hotfix right away), all you'd need (after this PR) is:
```
pkg "foo", "nimble test", url = "https://github.com/realauthor/foo"
=>
pkg "foo", "git checkout hotfix && nimble test", url = "https://github.com/hotfixauthor/foo"
```
and you don't have to rewrite the plubming to support seq[string] in important_packages.nim. (that was just an example; my point remains even if we do end up supporting seq[string] in this particular case)

in any case, we should strive for minimizing cross-platform differences, and now `execCmdEx("git checkout hotfix && nimble test")` works now cross platform

timotheecour committed:
fixup

Clyybber commented on testament/categories.nim:
```diff
@@ -101,7 +101,7 @@ proc flagTests(r: var TResults, cat: Category, options: string) =
 
   when defined(windows):
-    testExec r, makeTest(filename, " cmd /c cd " & nimcache &
+    testExec r, makeTest(filename, " cd " & nimcache &
```
Out of curiousity, why is this needed?

timotheecour replied:
in this particular case `cmd /c` would already be added by this PR's logic so I'm removing it from here; I am trying to fix CI failures; help welcome btw
* see https://dev.azure.com/nim-lang/255dfe86-e590-40bb-a8a2-3c0295ebdeb1/_apis/build/builds/2606/logs/68
* and https://dev.azure.com/nim-lang/255dfe86-e590-40bb-a8a2-3c0295ebdeb1/_apis/build/builds/2606/logs/72

Clyybber replied:
@timotheecour Sure, hop on gitter/irc?
